# YOUR FIRST LAB
### Please read this section. This is valuable to get you prepared, even if it's a long read -- it's important stuff.

### Also, be sure to read [README.md](../README.md)! More info about the updated videos in the README and [top of the course resources in purple](https://edwarddonner.com/2024/11/13/llm-engineering-resources/)

## Your first Frontier LLM Project

By the end of this course, you will have built an autonomous Agentic AI solution with 7 agents that collaborate to solve a business problem. All in good time! We will start with something smaller...

Our goal is to code a new kind of Web Browser. Give it a URL, and it will respond with a summary. The Reader's Digest of the internet!!

Before starting, you should have completed the setup linked in the README.

### If you're new to working in "Notebooks" (also known as Labs or Jupyter Lab)

Welcome to the wonderful world of Data Science experimentation! Simply click in each "cell" with code in it, such as the cell immediately below this text, and hit Shift+Return to execute that cell. Be sure to run every cell, starting at the top, in order.

Please look in the [Guides folder](../guides/01_intro.ipynb) for all the guides.

## I am here to help

If you have any problems at all, please do reach out.  
I'm available through the platform, or at ed@edwarddonner.com, or at https://www.linkedin.com/in/eddonner/ if you'd like to connect (and I love connecting!)  
And this is new to me, but I'm also trying out X at [@edwarddonner](https://x.com/edwarddonner) - if you're on X, please show me how it's done 😂  

## More troubleshooting

Please see the [troubleshooting](../setup/troubleshooting.ipynb) notebook in the setup folder to diagnose and fix common problems. At the very end of it is a diagnostics script with some useful debug info.

## If this is old hat!

If you're already comfortable with today's material, please hang in there; you can move swiftly through the first few labs - we will get much more in depth as the weeks progress. Ultimately we will fine-tune our own LLM to compete with OpenAI!

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Please read - important note</h2>
            <span style="color:#900;">The way I collaborate with you may be different to other courses you've taken. I prefer not to type code while you watch. Rather, I execute Jupyter Labs, like this, and give you an intuition for what's going on. My suggestion is that you carefully execute this yourself, <b>after</b> watching the lecture. Add print statements to understand what's going on, and then come up with your own variations. If you have a Github account, use this to showcase your variations. Not only is this essential practice, but it demonstrates your skills to others, including perhaps future clients or employers...</span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/resources.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#f71;">This code is a live resource - keep an eye out for my emails</h2>
            <span style="color:#f71;">I push updates to the code regularly. As people ask questions, I add more examples or improved commentary. As a result, you'll notice that the code below isn't identical to the videos. Everything from the videos is here; but I've also added better explanations and new models like DeepSeek. Consider this like an interactive book.<br/><br/>
                I try to send emails regularly with important updates related to the course. You can find this in the 'Announcements' section of Udemy in the left sidebar. You can also choose to receive my emails via your Notification Settings in Udemy. I'm respectful of your inbox and always try to add value with my emails!
            </span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business value of these exercises</h2>
            <span style="color:#181;">A final thought. While I've designed these notebooks to be educational, I've also tried to make them enjoyable. We'll do fun things like have LLMs tell jokes and argue with each other. But fundamentally, my goal is to teach skills you can apply in business. I'll explain business implications as we go, and it's worth keeping this in mind: as you build experience with models and techniques, think of ways you could put this into action at work today. Please do contact me if you'd like to discuss more or if you have ideas to bounce off me.</span>
        </td>
    </tr>
</table>

### If necessary, install Cursor Extensions

1. From the View menu, select Extensions
2. Search for Python
3. Click on "Python" made by "ms-python" and select Install if not already installed
4. Search for Jupyter
5. Click on "Jupyter" made by "ms-toolsai" and select Install if not already installed


### Next Select the Kernel

Click on "Select Kernel" on the Top Right

Choose "Python Environments..."

Then choose the one that looks like `.venv (Python 3.12.x) .venv/bin/python` - it should be marked as "Recommended" and have a big star next to it.

Any problems with this? Head over to the troubleshooting.

### Note: you'll need to set the Kernel with every notebook..

In [2]:
# imports

import os
from dotenv import load_dotenv
from scraper import fetch_website_contents
from IPython.display import Markdown, display
from openai import OpenAI

# If you get an error running this cell, then please head over to the troubleshooting notebook!

# Connecting to OpenAI (or Ollama)

The next cell is where we load in the environment variables in your `.env` file and connect to OpenAI.  

If you'd like to use free Ollama instead, please see the README section "Free Alternative to Paid APIs", and if you're not sure how to do this, there's a full solution in the solutions folder (day1_with_ollama.ipynb).

## Troubleshooting if you have problems:

If you get a "Name Error" - have you run all cells from the top down? Head over to the Python Foundations guide for a bulletproof way to find and fix all Name Errors.

If that doesn't fix it, head over to the [troubleshooting](../setup/troubleshooting.ipynb) notebook for step by step code to identify the root cause and fix it!

Or, contact me! Message me or email ed@edwarddonner.com and we will get this to work.

Any concerns about API costs? See my notes in the README - costs should be minimal, and you can control it at every point. You can also use Ollama as a free alternative, which we discuss during Day 2.

In [27]:
# Load environment variables in a file called .env

load_dotenv(override=True)
api_key = os.getenv('OPENAI_API_KEY')

# Check the key

if not api_key:
    print("No API key was found - please head over to the troubleshooting notebook in this folder to identify & fix!")
elif not api_key.startswith("sk-proj-"):
    print("An API key was found, but it doesn't start sk-proj-; please check you're using the right key - see troubleshooting notebook")
elif api_key.strip() != api_key:
    print("An API key was found, but it looks like it might have space or tab characters at the start or end - please remove them - see troubleshooting notebook")
else:
    print("API key found and looks good so far!")


API key found and looks good so far!


# Let's make a quick call to a Frontier model to get started, as a preview!

In [28]:
# To give you a preview -- calling OpenAI with these messages is this easy. Any problems, head over to the Troubleshooting notebook.

message = "Hello, GPT! This is my first ever message to you! Hi!"

messages = [{"role": "user", "content": message}]

messages


[{'role': 'user',
  'content': 'Hello, GPT! This is my first ever message to you! Hi!'}]

In [29]:
openai = OpenAI()

response = openai.chat.completions.create(model="gpt-5-nano", messages=messages)
response.choices[0].message.content

'Hi there! Welcome — nice to meet you. I’m ChatGPT, and I can help with questions, explanations, writing, planning, coding, brainstorming, and more.\n\nWhat would you like to do today? Here are some quick ideas:\n- Explain a topic in simple terms\n- Help write or edit a message, essay, or email\n- Brainstorm ideas for a project or story\n- Create a step-by-step plan or checklist\n- Write or debug code\n- Practice a language or study a subject\n- Do a quick quiz or riddle for fun\n\nIf you’d like, tell me your name and a bit about what you’re into. How can I help you right now?'

## OK onwards with our first project

In [30]:
# Let's try out this utility

ed = fetch_website_contents("https://edwarddonner.com")
print(ed)

Home - Edward Donner

Home
Connect Four
Outsmart
An arena that pits LLMs against each other in a battle of diplomacy and deviousness
About
Posts
Well, hi there.
I’m Ed. I like writing code and experimenting with LLMs, and hopefully you’re here because you do too. I also enjoy DJing (but I’m badly out of practice), amateur electronic music production (
very
amateur) and losing myself in
Hacker News
, nodding my head sagely to things I only half understand.
I’m the co-founder and CTO of
Nebula.io
. We’re applying AI to a field where it can make a massive, positive impact: helping people discover their potential and pursue their reason for being. Recruiters use our product today to source, understand, engage and manage talent. I’m previously the founder and CEO of AI startup untapt,
acquired in 2021
.
We work with groundbreaking, proprietary LLMs verticalized for talent, we’ve
patented
our matching model, and our award-winning platform has happy customers and tons of press coverage.
Conne

## Types of prompts

You may know this already - but if not, you will get very familiar with it!

Models like GPT have been trained to receive instructions in a particular way.

They expect to receive:

**A system prompt** that tells them what task they are performing and what tone they should use

**A user prompt** -- the conversation starter that they should reply to

In [31]:
# Define our system prompt - you can experiment with this later, changing the last sentence to 'Respond in markdown in Spanish."

system_prompt = """
You are a snarky assistant that analyzes the contents of a website,
and provides a short, snarky, humorous summary, ignoring text that might be navigation related.
Respond in markdown. Do not wrap the markdown in a code block - respond just with the markdown.
"""

In [32]:
# Define our user prompt

user_prompt_prefix = """
Here are the contents of a website.
Provide a short summary of this website.
If it includes news or announcements, then summarize these too.

"""

## Messages

The API from OpenAI expects to receive messages in a particular structure.
Many of the other APIs share this structure:

```python
[
    {"role": "system", "content": "system message goes here"},
    {"role": "user", "content": "user message goes here"}
]
```
To give you a preview, the next 2 cells make a rather simple call - we won't stretch the mighty GPT (yet!)

In [33]:
messages = [
    {"role": "system", "content": "You are a helpful assistant"},
    {"role": "user", "content": "What is 2 + 2?"}
]

response = openai.chat.completions.create(model="gpt-4.1-nano", messages=messages)
response.choices[0].message.content

'2 + 2 equals 4.'

## And now let's build useful messages for GPT-4.1-mini, using a function

In [34]:
# See how this function creates exactly the format above

def messages_for(website):
    return [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_prompt_prefix + website}
    ]

In [35]:
# Try this out, and then try for a few more websites

messages_for(ed)

[{'role': 'system',
  'content': '\nYou are a snarkyassistant that analyzes the contents of a website,\nand provides a short, snarky, humorous summary, ignoring text that might be navigation related.\nRespond in markdown. Do not wrap the markdown in a code block - respond just with the markdown.\n'},
 {'role': 'user',
  'content': '\nHere are the contents of a website.\nProvide a short summary of this website.\nIf it includes news or announcements, then summarize these too.\n\nHome - Edward Donner\n\nHome\nConnect Four\nOutsmart\nAn arena that pits LLMs against each other in a battle of diplomacy and deviousness\nAbout\nPosts\nWell, hi there.\nI’m Ed. I like writing code and experimenting with LLMs, and hopefully you’re here because you do too. I also enjoy DJing (but I’m badly out of practice), amateur electronic music production (\nvery\namateur) and losing myself in\nHacker News\n, nodding my head sagely to things I only half understand.\nI’m the co-founder and CTO of\nNebula.io\n. 

## Time to bring it together - the API for OpenAI is very simple!

In [36]:
# And now: call the OpenAI API. You will get very familiar with this!
#openai = OpenAI()
def summarize(url):
    website = fetch_website_contents(url)
    response = openai.chat.completions.create(
        model = "gpt-4.1-mini",
        messages = messages_for(website)
    )
    return response.choices[0].message.content

In [37]:
summarize("https://edwarddonner.com")

'# Edward Donner’s Website: The Snarky Edition\n\nSo, Ed is a code-loving LLM tamer who moonlights as a (very) amateur electronic music enthusiast and occasional nodder of things he kinda gets on Hacker News. He\'s big on AI, running a startup called Nebula.io that helps recruiters find talent with some fancy proprietary LLMs and a patented matching model because why not patent matchmaking? Before that, he sold another AI startup, because apparently, that’s just what you do.\n\nHighlight: Ed hosts an epic LLM smackdown called "Outsmart," where AI chatbots duke it out in battles of diplomacy and deviousness. Sounds like the AI version of a Game of Thrones reunion.\n\nRecent news? Ed’s been dropping knowledge bombs with talks about AI live events, AI in production on AWS, and AI leadership curricula—because keeping humans in the loop is so 2025.\n\nWant to geek out? Connect with him or subscribe to his newsletter to stay in the loop or just marvel at his blend of tech and tunes.'

In [38]:
# A function to display this nicely in the output, using markdown

def display_summary(url):
    summary = summarize(url)
    display(Markdown(summary))

In [39]:
display_summary("https://edwarddonner.com")

PermissionDeniedError: Error code: 403 - {'error': {'message': 'Project `proj_tm1mIX8gThQguVffmElBMyTn` does not have access to model `gpt-4.1-mini`', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_found'}}

# Let's try more websites

Note that this will only work on websites that can be scraped using this simplistic approach.

Websites that are rendered with Javascript, like React apps, won't show up. See the community-contributions folder for a Selenium implementation that gets around this. You'll need to read up on installing Selenium (ask ChatGPT!)

Also Websites protected with CloudFront (and similar) may give 403 errors - many thanks Andy J for pointing this out.

But many websites will work just fine!

In [ ]:
display_summary("https://cnn.com")

In [ ]:
display_summary("https://anthropic.com")

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business applications</h2>
            <span style="color:#181;">In this exercise, you experienced calling the Cloud API of a Frontier Model (a leading model at the frontier of AI) for the first time. We will be using APIs like OpenAI at many stages in the course, in addition to building our own LLMs.

More specifically, we've applied this to Summarization - a classic Gen AI use case to make a summary. This can be applied to any business vertical - summarizing the news, summarizing financial performance, summarizing a resume in a cover letter - the applications are limitless. Consider how you could apply Summarization in your business, and try prototyping a solution.</span>
        </td>
    </tr>
</table>

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Before you continue - now try yourself</h2>
            <span style="color:#900;">Use the cell below to make your own simple commercial example. Stick with the summarization use case for now. Here's an idea: write something that will take the contents of an email, and will suggest an appropriate short subject line for the email. That's the kind of feature that might be built into a commercial email tool.</span>
        </td>
    </tr>
</table>

In [8]:
#!uv add tavily

Resolved 267 packages in 342ms                                       
Prepared 1 package in 263ms                                              ⠋ Preparing packages... (0/0)                                                   
Uninstalled 1 package in 41ms
Installed 2 packages in 12ms                                
 - pip==24.3.1 (from file:///Users/darthjedi/Applications/PyCharm.app/Contents/plugins/python-ce/helpers/pip-24.3.1-py2.py3-none-any.whl)
 + pip==25.3
 + tavily==1.1.0


In [19]:
import os
from dotenv import load_dotenv
from IPython.display import Markdown, display
from scraper import fetch_website_contents
from openai import OpenAI

from tavily import TavilyClient

load_dotenv(override=True)

def messages_for(website):
    return [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_prompt_prefix + website}
    ]

# Step 1: Create your prompts
openai_client = OpenAI()
system_prompt = """You are a sociologist who studies people's blogs, and then describes
    details about the person who wrote the blog. When you are given the text from a blog,
    determine who the person is, and then describe some of their top personality traits."""
user_prompt_prefix = "Read this blog and tell me what you know about this person:\n\n"
#website_content = fetch_website_contents("https://darthjedi.logiodice.com")
website_content = TavilyClient().extract(
    [
        "https://darthjedi.logiodice.com/?paged=1&cat=7",
        "https://darthjedi.logiodice.com/?paged=2&cat=7",
        "https://darthjedi.logiodice.com/?paged=3&cat=7",
        "https://darthjedi.logiodice.com/?paged=4&cat=7",
        ]
)
# Step 2: Make the messages list
raw_messages = "".join(x["raw_content"] for x in website_content["results"])
message = messages_for(raw_messages)

# Step 3: Call OpenAI
# response =
response = openai_client.chat.completions.create(
    model = "gpt-5",
    messages = message
)
content_response = response.choices[0].message.content
display(Markdown(content_response))

# Step 4: print the result
# print(

Who this appears to be
- Name: Jediah Logiodice
- Roles: Christian husband and father of several children (including adoptees), full-time professional in tech/cybersecurity (he jokes about “breaking into computers”), and a current doctoral student (visited Liberty University and got a student ID).
- Background: Grew up very poor without running water, Italian heritage, active in church leadership (missions board), has officiated weddings.
- Family: Deeply family-centered; mentions wife Amanda, multiple kids at driving age, a married middle son (2024), adopted children (including from Haiti). Pets abound.
- Interests: Faith/ministry, adoption/orphan care, motorcycles, accordions/music, DIY projects, camping/nature, films, health/nutrition (Keto), pop culture.

What their life looks like
- Balances full-time work, doctoral studies, and a large, busy household.
- Strong evangelical Christian identity; frequent faith reflections; attends/engages with Liberty University events; expresses “God is good – all the time.”
- Adoption is central to family life and theology; aware of trauma and complexity and speaks about “being chosen.”
- Tech-savvy, likely cybersecurity/IT; comfortable mentoring kids in tech.
- Enjoys hands-on and outdoorsy pursuits (motorcycles, camping, building projects), and is musically inclined (plays accordion/Cordovox).

Top personality traits (with signals)
- Devout and values-driven: Frames life through faith; served on a missions board; adopts as an expression of theology; “In Excelsis Deo!”
- Family-first and nurturing: Frequent posts about kids’ milestones, college tours, driving lessons; empathetic about adoptees’ grief.
- Resilient and hard-working: Juggles work, doctoral studies, and parenting; “surprise myself by each and every year I cross the finish line.”
- Compassionate and service-oriented: Advocates for orphans/widows, commends films that spur action; recounts “God moments” in adoption.
- Honest and unvarnished: Pushes back on social-media perfection; openly describes the messiness of parenting and marriage humor.
- Playful, self-deprecating humor: Teases about weight, chiropractors, cats, and gift-giving; names his bike “Bob”; uses pop-culture quips.
- Independent thinker/nonconformist: Encourages kids to “be themselves”; questions dietary guidelines; wore a homemade tie to church to model individuality.
- Lifelong learner/mentor: Pursues a doctorate; brings kids into his interests (tech, music, motorcycles).
- Techie/tinkerer: References penetration testing; builds things (fish stand); comfortable with gadgets and remote-work life.
- Community-minded yet pragmatic: Engages local issues (BLM protests) with a tone of calm, unity, and practical concern for nonviolence.

Overall impression: A faith-anchored, hands-on, tech-savvy dad who blends service, humor, and grit—invested in his family’s growth, unafraid to be candid about the hard parts, and committed to learning and leading by example.

## An extra exercise for those who enjoy web scraping

You may notice that if you try `display_summary("https://openai.com")` - it doesn't work! That's because OpenAI has a fancy website that uses Javascript. There are many ways around this that some of you might be familiar with. For example, Selenium is a hugely popular framework that runs a browser behind the scenes, renders the page, and allows you to query it. If you have experience with Selenium, Playwright or similar, then feel free to improve the Website class to use them. In the community-contributions folder, you'll find an example Selenium solution from a student (thank you!)

# Sharing your code

I'd love it if you share your code afterwards so I can share it with others! You'll notice that some students have already made changes (including a Selenium implementation) which you will find in the community-contributions folder. If you'd like add your changes to that folder, submit a Pull Request with your new versions in that folder and I'll merge your changes.

If you're not an expert with git (and I am not!) then GPT has given some nice instructions on how to submit a Pull Request. It's a bit of an involved process, but once you've done it once it's pretty clear. As a pro-tip: it's best if you clear the outputs of your Jupyter notebooks (Edit >> Clean outputs of all cells, and then Save) for clean notebooks.

Here are good instructions courtesy of an AI friend:  
https://chatgpt.com/share/677a9cb5-c64c-8012-99e0-e06e88afd293